File: Library/OldCode/smog-swift-v1/smog/interpreter/Interpreter.swift
//
//  Interpreter.swift
//  smog
//
//  Created by Kristofer Younger on 8/17/23.
//

import Foundation

class Interpreter {
    
    var u: Universe
    var frame: Frame?
    
    init(_ u: Universe) {
        self.u = u
        self.frame = nil //Universe.shared.nilObject
    }
    
    //  doDup = (
    //    frame push: (frame stackElement: 0)
    //  )
    func doDup() {
        if let frame = self.frame {
            frame.push(obj: (frame.stackElement(idx: 0)))
        }
    }
    //  doPushLocal: bytecodeIndex = (
    //    frame push: (
    //        frame local: (frame method bytecode: bytecodeIndex + 1)
    //                 at: (frame method bytecode: bytecodeIndex + 2))
    //  )
    func doPushLocal(idx: Int) {
        if let frame = self.frame {
            frame.push(
                obj: (frame.local(idx: (frame.method.bytecode(at: idx + 1)),
                                  at: (frame.method.bytecode(at: idx + 2)))))
        }
    }
    //
    //  doPushArgument: bytecodeIndex = (
    //    frame push: (
    //        frame argument: (frame method bytecode: bytecodeIndex + 1)
    //                    at: (frame method bytecode: bytecodeIndex + 2))
    //  )
    func doPushArgument(idx: Int) {
        if let frame = self.frame {
            frame.push(
                obj: (frame.argument(idx: (frame.method.bytecode(at: idx + 1)),
                                     at: (frame.method.bytecode(at: idx + 2)))))
        }
    }
    //
    //  doPushField: bytecodeIndex = (
    //    | fieldIndex |
    //    fieldIndex := frame method bytecode: bytecodeIndex + 1.
    //
    //    "Push the field with the computed index onto the stack"
    //    frame push: (self getSelf field: fieldIndex)
    //  )
    func doPushField(idx: Int) {
        if let frame = self.frame {
            let fieldIndex = frame.method.bytecode(at: idx + 1)
            
            frame.push(obj: self.getSelf().field(index: fieldIndex))
        }
    }
    
    //
    //  doPushBlock: bytecodeIndex = (
    //    | blockMethod |
    //    blockMethod := frame method constant: bytecodeIndex.
    //
    //    "Push a new block with the current frame as context onto the stack"
    //    frame push: (
    //        universe newBlock: blockMethod
    //                     with: frame
    //                  numArgs: blockMethod numberOfArguments)
    //  )
    func doPushBlock(idx: Int) {
        if let frame = self.frame {
            let blockMethod = frame.method.constant(bcIndex: idx) as! SMethod
            let univ = Universe.shared
            // "Push a new block with the current frame as context onto the stack"
            
            frame.push(
                obj: univ.newBlock(method: blockMethod , with: frame, numArgs: blockMethod.numberOfArguments())
            )
        }
    }
    
    //
    //  doPushConstant: bytecodeIndex = (
    //    frame push: (frame method constant: bytecodeIndex)
    //  )
    func doPushConstant(idx: Int) {
        if let frame = self.frame {
            frame.push(obj: frame.method.constant(bcIndex: idx))
        }
    }
    
    //
    //  doPushGlobal: bytecodeIndex = (
    //    | globalName global |
    //    globalName := frame method constant: bytecodeIndex.
    //
    //    "Get the global from the universe"
    //    global := universe global: globalName.
    //
    //    global ~= nil
    //      ifTrue: [ frame push: global  ]
    //      ifFalse: [
    //        "Send 'unknownGlobal:' to self"
    //        self getSelf sendUnknownGlobal: globalName in: universe using: self ]
    //  )
    func doPushGlobal(idx: Int) {
        if let frame = self.frame {
            let globalName = frame.method.constant(bcIndex: idx)
            let global = Universe.shared.global(globalName as! SSymbol)
            if global != Universe.shared.nilObject {
                frame.push(obj: global)
            } else {
                //        "Send 'unknownGlobal:' to self"
                let ns = globalName as! SSymbol
                self.getSelf().sendUnknownGlobal(ns.s, in: Universe.shared, using: self)
            }
        }
    }
    
    //
    //  doPop = (
    //    frame pop
    //  )
    func doPop() {
        if let frame = self.frame {
            _ = frame.pop()
        }
    }
    //
    //  doPopLocal: bytecodeIndex = (
    //    frame local: (frame method bytecode: bytecodeIndex + 1)
    //             at: (frame method bytecode: bytecodeIndex + 2)
    //            put: frame pop
    //  )
    func doPopLocal(idx: Int) {
        if let frame = self.frame {
            frame.local(idx: frame.method.bytecode(at: idx + 1),
                        at: frame.method.bytecode(at: idx + 2),
                        put: frame.pop())
        }
    }
    
    //
    //  doPopArgument: bytecodeIndex = (
    //    frame argument: (frame method bytecode: bytecodeIndex + 1)
    //                at: (frame method bytecode: bytecodeIndex + 2)
    //               put: frame pop
    //  )
    func doPopArgument(idx: Int) {
        if let frame = self.frame {
            frame.argument(idx: frame.method.bytecode(at: idx + 1),
                           at: frame.method.bytecode(at: idx + 2),
                           put: frame.pop())
        }
    }
    
    //
    //  doPopField: bytecodeIndex = (
    //    | fieldIndex |
    //    fieldIndex := frame method bytecode: bytecodeIndex + 1.
    //
    //    "Set the field with the computed index to the value popped from the stack"
    //    self getSelf field: fieldIndex put: frame pop
    //  )
    func doPopField(idx: Int) {
        if let frame = self.frame {
            let fieldIndex = frame.method.bytecode(at: idx + 1)
            //    "Set the field with the computed index to the value popped from the stack"
            self.getSelf().fieldAt(idx, put: frame.pop())
        }
    }
    
    //
    //  doSend: bytecodeIndex = (
    //    | signature numberOfArguments receiver |
    //    signature := frame method constant: bytecodeIndex.
    //    numberOfArguments := signature numberOfSignatureArguments.
    //    receiver := frame stackElement: numberOfArguments - 1.
    //    self send: signature rcvrClass: (receiver somClassIn: universe)
    //  )
    func doSend(idx: Int) {
        if let frame = self.frame {
            let signature = frame.method.constant(bcIndex: idx) as! SSymbol
            let numArgs = signature.numSignatureArguments
            let receiver = frame.stackElement(idx: numArgs - 1)
            self.send(selector: signature, rcvrClass: receiver.somClass())
        }
    }
    //
    //  doSuperSend: bytecodeIndex = (
    //    | signature holderSuper invokable |
    //    signature := frame method constant: bytecodeIndex.
    //
    //    "Send the message
    //     Lookup the invokable with the given signature"
    //    holderSuper := frame method holder superClass.
    //    invokable := holderSuper lookupInvokable: signature.
    //
    //    self activate: invokable orDnu: signature
    //  )
    func doSuperSend(idx: Int) {
        if let frame = self.frame {
            let signature = frame.method.constant(bcIndex: idx) as! SSymbol
            //    "Send the message
            //     Lookup the invokable with the given signature"
            let holderSuper = frame.method.holderClass.clazz.superClass
            let invokable = holderSuper.lookupInvokable(signature: signature)
            
            self.activate(invokable: invokable, orDnu: signature)
        }
    }
    //
    //  doReturnLocal = (
    //    | result |
    //    result := frame pop.
    //
    //    "Pop the top frame and push the result"
    //    self popFrameAndPushResult: result
    //  )
    func doReturnLocal() {
        if let frame = self.frame {
            let result = frame.pop()
            self.popFrameAndPushResult(result: result)
        }
    }
    //
    //  doReturnNonLocal = (
    //    | result context |
    //    result := frame pop.
    //
    //    "Compute the context for the non-local return"
    //    context := frame outerContext.
    //
    //    "Make sure the block context is still on the stack"
    //    context hasPreviousFrame ifFalse: [
    //      | block sender method numArgs |
    //      "Try to recover by sending 'escapedBlock:' to the sending object
    //       this can get a bit nasty when using nested blocks. In this case
    //       the 'sender' will be the surrounding block and not the object
    //       that actually sent the 'value' message."
    //      block := frame argument: 1 at: 0.
    //      sender := frame previousFrame outerContext argument: 1 at: 0.
    //
    //      "pop the frame of the currently executing block..."
    //      self popFrame.
    //
    //      "pop old arguments from stack"
    //      method := frame method.
    //      numArgs := method numberOfArguments.
    //      numArgs timesRepeat: [ frame pop ].
    //
    //      "... and execute the escapedBlock message instead"
    //      sender sendEscapedBlock: block in: universe using: self.
    //      ^ self ].
    //
    //    "Unwind the frames"
    //    [frame ~= context] whileTrue: [
    //      self popFrame ].
    //
    //    self popFrameAndPushResult: result
    //  )
    func doReturnNonLocal() {
        if let frame = self.frame {
            let result = frame.pop()
            let context = frame.outerContext()
            
            if context.hasPreviousFrame() == false {
                // "Try to recover by sending 'escapedBlock:' to the sending object
                // this can get a bit nasty when using nested blocks. In this case
                // the 'sender' will be the surrounding block and not the object
                // that actually sent the 'value' message."
                let block = frame.argument(idx: 1, at: 0) as! SBlock
                let sender = frame.previousFrame?.outerContext().argument(idx: 1, at: 0)
                
                _ = self.popFrame()
                
                let method = frame.method
                let numArgs = method.numberOfArguments()
                for _ in 0...numArgs { _ = frame.pop() }
                //      "... and execute the escapedBlock message instead"
                //      sender sendEscapedBlock: block in: universe using: self.
                sender!.sendEscapedBlock(block, in: Universe.shared, using: self)
                //return self
            }
            //    "Unwind the frames"
            //    [frame ~= context] whileTrue: [
            //      self popFrame ].
            while !(frame == context) {
                _ = self.popFrame()
            }
            self.popFrameAndPushResult(result: result)
        }
    }
    
    //
    //  start = (
    //    [true] whileTrue: [
    //      | bytecodeIndex bytecode bytecodeLength nextBytecodeIndex result |
    //      bytecodeIndex := frame bytecodeIndex.
    //      bytecode := frame method bytecode: bytecodeIndex.
    //      bytecodeLength := Bytecodes length: bytecode.
    //      nextBytecodeIndex := bytecodeIndex + bytecodeLength.
    //      frame bytecodeIndex: nextBytecodeIndex.
    //
    //      result := self dispatch: bytecode idx: bytecodeIndex.
    //      result ~= nil
    //        ifTrue: [ ^ result ] ]
    //  )
    
    func start() -> SObject {
        while true {
            let bytecodeIndex = self.frame!.bytecodeIndex
            let bytecode = self.frame!.method.bytecode(at: bytecodeIndex)
            
            let nextBytecodeIndex = bytecodeIndex + bytecodeLength // len is global
            self.frame!.bytecodeIndex(idx: nextBytecodeIndex)
            let result = self.dispatch(bc: bytecode, bytecodeIndex: bytecodeIndex)
            if !(result == Universe.shared.nilObject) {
                return result
            }
            
        }
    }
    
    //  dispatch: bytecode idx: bytecodeIndex = (
    func dispatch(bc: Int, bytecodeIndex: Int) -> SObject {
        let nilObject = Universe.shared.nilObject
        let bytecode = Bc(rawValue: bc)
        if bytecode == Bc.halt { return (frame?.stackElement(idx: bytecodeIndex))! }
        if bytecode == Bc.dup {
            self.doDup()
            return nilObject
        }
        if bytecode == Bc.pushLocal {
            self.doPushLocal(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.pushArgument {
            self.doPushArgument(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.pushField {
            self.doPushField(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.pushBlock {
            self.doPushBlock(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.pushConstant {
            self.doPushConstant(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.pushGlobal {
            self.doPushGlobal(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.pop {
            self.doPop()
            return nilObject
        }
        if bytecode == Bc.popLocal {
            self.doPopLocal(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.popArgument {
            self.doPopArgument(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.popField {
            self.doPopField(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.send {
            self.doSend(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.superSend {
            self.doSuperSend(idx: bytecodeIndex)
            return nilObject
        }
        if bytecode == Bc.returnLocal {
            self.doReturnLocal()
            return nilObject
        }
        if bytecode == Bc.returnNonLocal {
            self.doReturnNonLocal()
            return nilObject
        }
        //    self error: 'Unknown bytecode' + bytecode asString
        print("Unknown bytecode \(bytecode!.rawValue)")
        return Universe.shared.nilObject
    }
    //
    //  pushNewFrame: method with: contextFrame = (
    //    frame := universe newFrame: frame with: method with: contextFrame.
    //    ^ frame
    //  )
    func pushNewFrame(invokable: Invokable, withContextFrame: Frame?) -> Frame {
        let f = Universe.shared.newFrame(previousFrame: self.frame!, method: invokable, withContextFrame: withContextFrame)
        return f
    }
    //
    //  pushNewFrame: method = (
    //    ^ self pushNewFrame: method with: nil
    //  )
    func pushNewFrame(invokable: Invokable) -> Frame {
        self.pushNewFrame(invokable: invokable, withContextFrame: nil)
    }
    
    //
    //  frame = (
    //    ^ frame
    //  )
    //
    //  method = (
    //    ^ frame method
    //  )
    //
    //  getSelf = (
    //    "Get the self object from the interpreter"
    //    ^ frame outerContext argument: 1 at: 0
    //  )
    func getSelf() -> SObject {
        //    "Get the self object from the interpreter"
        if let frame = self.frame {
            
            return frame.outerContext().argument(idx: 1, at: 0)
        }
        return Universe.shared.nilObject
    }
    //
    //  send: selector rcvrClass: receiverClass = (
    //    | invokable |
    //    invokable := receiverClass lookupInvokable: selector.
    //    self activate: invokable orDnu: selector
    //  )
    func send(selector: SSymbol, rcvrClass: SClass) {
        let invokable = rcvrClass.lookupInvokable(signature: selector)
        self.activate(invokable: invokable, orDnu: selector)
    }
    //
    //  activate: invokable orDnu: signature = (
    //    invokable ~= nil
    //        ifTrue: [
    //          "Invoke the invokable in the current frame"
    //          invokable invoke: frame using: self ]
    //        ifFalse: [
    //          | numberOfArguments receiver |
    //          numberOfArguments := signature numberOfSignatureArguments.
    //          receiver := frame stackElement: numberOfArguments - 1.
    //          receiver sendDoesNotUnderstand: signature in: universe using: self ]
    //  )
    func activate(invokable: Invokable, orDnu signature: SSymbol) {
        if let frame = self.frame {
            if invokable != nil {
                invokable.invoke(frame: frame, using: self )
            } else {
                let numberOfArguments = signature.numSignatureArguments
                let receiver = frame.stackElement(idx: numberOfArguments - 1)
                receiver.sendDoesNotUnderstand(signature.string(), in: self.u, using: self)
            }
        }
    }
    //
    //  popFrame = (
    //    | result |
    //    "Save a reference to the top frame"
    //    result := frame.
    //
    //    "Pop the top frame from the frame stack"
    //    frame := frame previousFrame.
    //
    //    "Destroy the previous pointer on the old top frame"
    //    result clearPreviousFrame.
    //
    //    "Return the popped frame"
    //    ^ result
    //  )
    func popFrame() -> Frame {
        let result = self.frame
        self.frame = self.frame!.previousFrame
        result?.clearPreviousFrame()
        return result!
    }
    //
    //  popFrameAndPushResult: result = (
    //    | numberOfArguments |
    //    "Pop the top frame from the interpreter frame stack and
    //     get the number of arguments"
    //    numberOfArguments := self popFrame method numberOfArguments.
    //
    //    "Pop the arguments"
    //    numberOfArguments
    //      timesRepeat: [ frame pop ].
    //
    //    frame push: result
    //  )
    func popFrameAndPushResult(result: SObject) {
        //    "Pop the top frame from the interpreter frame stack and
        //     get the number of arguments and pop them"
        let numArgs = self.popFrame().method.numberOfArguments()
        for i in 0...numArgs {
            _ = self.frame!.pop()
        }
        self.frame?.push(obj: result)
    }
    //
}

File: src/internal/vm/interpreter.go
package vm

import (
	"fmt"
)

type Interpreter struct{}

func (i *Interpreter) doDup() {
	// Handle the dup bytecode
	frame := getFrame()
	frame.push(frame.getStackElement(0))
}

func (i *Interpreter) doPushLocal(bytecodeIndex int) {
	// Handle the push local bytecode
	frame := getFrame()
	frame.push(frame.getLocal(getMethod().getBytecode(bytecodeIndex+1), getMethod().getBytecode(bytecodeIndex+2)))
}

func (i *Interpreter) doPushArgument(bytecodeIndex int) {
	// Handle the push argument bytecode
	frame := getFrame()
	frame.push(frame.getArgument(getMethod().getBytecode(bytecodeIndex+1), getMethod().getBytecode(bytecodeIndex+2)))
}

func (i *Interpreter) doPushField(bytecodeIndex int) {
	// Handle the push field bytecode
	fieldName := getMethod().getConstant(bytecodeIndex).(Symbol)

	// Get the field index from the field name
	fieldIndex := getSelf().getFieldIndex(fieldName)

	// Push the field with the computed index onto the stack
	frame := getFrame()
	frame.push(getSelf().getField(fieldIndex))
}

func (i *Interpreter) doPushBlock(bytecodeIndex int) {
	// Handle the push block bytecode
	blockMethod := getMethod().getConstant(bytecodeIndex).(Method)

	// Push a new block with the current frame as context onto the stack
	frame := getFrame()
	frame.push(Universe.newBlock(blockMethod, frame, blockMethod.getNumberOfArguments()))
}

func (i *Interpreter) doPushConstant(bytecodeIndex int) {
	// Handle the push constant bytecode
	frame := getFrame()
	frame.push(getMethod().getConstant(bytecodeIndex))
}

func (i *Interpreter) doPushGlobal(bytecodeIndex int) {
	// Handle the push global bytecode
	globalName := getMethod().getConstant(bytecodeIndex).(Symbol)

	// Get the global from the universe
	global := Universe.getGlobal(globalName)

	if global != nil {
		// Push the global onto the stack
		frame := getFrame()
		frame.push(global)
	} else {
		// Send 'unknownGlobal:' to self
		arguments := []interface{}{globalName}
		getSelf().send("unknownGlobal:", arguments)
	}
}

func (i *Interpreter) doPop() {
	// Handle the pop bytecode
	frame := getFrame()
	frame.pop()
}

func (i *Interpreter) doPopLocal(bytecodeIndex int) {
	// Handle the pop local bytecode
	frame := getFrame()
	frame.setLocal(getMethod().getBytecode(bytecodeIndex+1), getMethod().getBytecode(bytecodeIndex+2), frame.pop())
}

func (i *Interpreter) doPopArgument(bytecodeIndex int) {
	// Handle the pop argument bytecode
	frame := getFrame()
	frame.setArgument(getMethod().getBytecode(bytecodeIndex+1), getMethod().getBytecode(bytecodeIndex+2), frame.pop())
}

func (i *Interpreter) doPopField(bytecodeIndex int) {
	// Handle the pop field bytecode
	fieldName := getMethod().getConstant(bytecodeIndex).(Symbol)

	// Get the field index from the field name
	fieldIndex := getSelf().getFieldIndex(fieldName)

	// Set the field with the computed index to the value popped from the stack
	getSelf().setField(fieldIndex, getFrame().pop())
}

func (i *Interpreter) doSuperSend(bytecodeIndex int) {
	// Handle the super send bytecode
	signature := getMethod().getConstant(bytecodeIndex).(Symbol)

	// Send the message
	// Lookup the invokable with the given signature
	invokable := getMethod().getHolder().getSuperClass().lookupInvokable(signature)

	if invokable != nil {
		// Invoke the invokable in the current frame
		invokable.invoke(getFrame())

	} else {
		// Compute the number of arguments
		numberOfArguments := signature.getNumberOfSignatureArguments()

		// Compute the receiver
		receiver := getFrame().getStackElement(numberOfArguments - 1)

		// Allocate an array with enough room to hold all arguments
		argumentsArray := Universe.newArray(numberOfArguments)

		// Remove all arguments and put them in the freshly allocated array
		for i := numberOfArguments - 1; i >= 0; i-- {
			argumentsArray.setIndexableField(i, getFrame().pop())
		}

		// Send 'doesNotUnderstand:arguments:' to the receiver object
		arguments := []interface{}{signature, argumentsArray}
		receiver.send("doesNotUnderstand:arguments:", arguments)
	}
}

func (i *Interpreter) doReturnLocal() {
	// Handle the return local bytecode
	result := getFrame().pop()

	// Pop the top frame and push the result
	popFrameAndPushResult(result)
}

func (i *Interpreter) doReturnNonLocal() {
	// Handle the return non local bytecode
	result := getFrame().pop()

	// Compute the context for the non-local return
	context := getFrame().getOuterContext()

	// Make sure the block context is still on the stack
	if !context.hasPreviousFrame() {
		// Try to recover by sending 'escapedBlock:' to the sending object
		// this can get a bit nasty when using nested blocks. In this case
		// the "sender" will be the surrounding block and not the object that
		// actually sent the 'value' message.
		block := getFrame().getArgument(0, 0).(Block)
		sender := getFrame().getPreviousFrame().getOuterContext().getArgument(0, 0)
		arguments := []interface{}{block}

		// pop the frame of the currently executing block...
		popFrame()

		// ... and execute the escapedBlock message instead
		sender.send("escapedBlock:", arguments)

		return
	}

	// Unwind the frames
	for getFrame() != context {
		popFrame()
	}

	// Pop the top frame and push the result
	popFrameAndPushResult(result)
}

func (i *Interpreter) doSend(bytecodeIndex int) {
	// Handle the send bytecode
	signature := getMethod().getConstant(bytecodeIndex).(Symbol)

	// Get the number of arguments from the signature
	numberOfArguments := signature.getNumberOfSignatureArguments()

	// Get the receiver from the stack
	receiver := getFrame().getStackElement(numberOfArguments - 1)

	// Send the message
	send(signature, receiver.getSOMClass(), bytecodeIndex)
}

func (i *Interpreter) start() {
	// Iterate through the bytecodes
	for {
		// Get the current bytecode index
		bytecodeIndex := getFrame().getBytecodeIndex()

		// Get the current bytecode
		bytecode := getMethod().getBytecode(bytecodeIndex)

		// Get the length of the current bytecode
		bytecodeLength := getBytecodeLength(bytecode)

		// Compute the next bytecode index
		nextBytecodeIndex := bytecodeIndex + bytecodeLength

		// Update the bytecode index of the frame
		getFrame().setBytecodeIndex(nextBytecodeIndex)

		// Handle the current bytecode
		switch bytecode {

		case HALT:
			{
				// Handle the halt bytecode
				return
			}

		case DUP:
			{
				i.doDup()
				break
			}

		case PUSH_LOCAL:
			{
				i.doPushLocal(bytecodeIndex)
				break
			}

		case PUSH_ARGUMENT:
			{
				i.doPushArgument(bytecodeIndex)
				break
			}

		case PUSH_FIELD:
			{
				i.doPushField(bytecodeIndex)
				break
			}

		case PUSH_BLOCK:
			{
				i.doPushBlock(bytecodeIndex)
				break
			}

		case PUSH_CONSTANT:
			{
				i.doPushConstant(bytecodeIndex)
				break
			}

		case PUSH_GLOBAL:
			{
				i.doPushGlobal(bytecodeIndex)
				break
			}

		case POP:
			{
				i.doPop()
				break
			}

		case POP_LOCAL:
			{
				i.doPopLocal(bytecodeIndex)
				break
			}

		case POP_ARGUMENT:
			{
				i.doPopArgument(bytecodeIndex)
				break
			}

		case POP_FIELD:
			{
				i.doPopField(bytecodeIndex)
				break
			}

		case SEND:
			{
				i.doSend(bytecodeIndex)
				break
			}

		case SUPER_SEND:
			{
				i.doSuperSend(bytecodeIndex)
				break
			}

		case RETURN_LOCAL:
			{
				i.doReturnLocal()
				break
			}

		case RETURN_NON_LOCAL:
			{
				i.doReturnNonLocal()
				break
			}

		default:
			fmt.Println("Nasty bug in interpreter")
			break
		}
	}
}

func (i *Interpreter) pushNewFrame(method Method) Frame {
	// Allocate a new frame and make it the current one
	frame := Universe.newFrame(frame, method)

	// Return the freshly allocated and pushed frame
	return frame
}

func getFrame() Frame {
	// Get the frame from the interpreter
	return frame
}

func getMethod() Method {
	// Get the method from the interpreter
	return getFrame().getMethod()
}

func getSelf() Object {
	// Get the self object from the interpreter
	return getFrame().getOuterContext().getArgument(0, 0)
}

func send(signature Symbol, receiverClass Class, bytecodeIndex int) {
	// Lookup the invokable with the given signature
	invokable := receiverClass.lookupInvokable(signature)

	if invokable != nil {
		// Invoke the invokable in the current frame
		invokable.invoke(getFrame())

	} else {
		// Compute the number of arguments
		numberOfArguments := signature.getNumberOfSignatureArguments()

		// Compute the receiver
		receiver := getFrame().getStackElement(numberOfArguments - 1)

		// Allocate an array with enough room to hold all arguments
		argumentsArray := Universe.newArray(numberOfArguments)

		// Remove all arguments and put them in the freshly allocated array
		for i := numberOfArguments - 1; i >= 0; i-- {
			argumentsArray.setIndexableField(i, getFrame().pop())
		}

		// Send 'doesNotUnderstand:arguments:' to the receiver object
		arguments := []interface{}{signature, argumentsArray}
		receiver.send("doesNotUnderstand:arguments:", arguments)
	}
}

func popFrame() Frame {
	// Save a reference to the top frame
	result := frame

	// Pop the top frame from the frame stack
	frame = frame.getPreviousFrame()

	// Destroy the previous pointer on the old top frame
	result.clearPreviousFrame()

	// Return the popped frame
	return result
}

func popFrameAndPushResult(result Object) {
	// Pop the top frame from the interpreter frame stack and compute the number of arguments
	numberOfArguments := popFrame().getMethod().getNumberOfArguments()

	// Pop the arguments
	for i := 0; i < numberOfArguments; i++ {
		getFrame().pop()
	}

	// Push the result
	getFrame().push(result)
}

var frame Frame

func main() {
	interpreter := &Interpreter{}
	interpreter.start()
}
